src/wallpaper.c

```diff
@@ -0,0 +1,429 @@
+
+          id = g_strconcat (app_id, ".desktop", NULL);
+          info = g_desktop_app_info_new (id);
+          name = g_app_info_get_display_name (G_APP_INFO (info));
+
+          title = g_strdup_printf (_("Allow %s to Set Backgrounds?"), name);
+          subtitle = g_strdup_printf (_("%s is requesting to be able to change the background image."), name);
+        }
+
+      body = _("This permission can be changed at any time from the privacy settings.");
+
+      if (!xdp_impl_access_call_access_dialog_sync (access_impl,
+                                                    request->id,
+                                                    app_id,
+                                                    parent_window,
+                                                    title,
+                                                    subtitle,
+                                                    body,
+                                                    g_variant_builder_end (&access_opt_builder),
+                                                    &access_response,
+                                                    &access_results,
+                                                    NULL,
+                                                    &error))
+        {
+          g_warning ("Failed to show access dialog: %s", error->message);
+          return;
+        }
+
+      if (access_response == 0)
+        {
+          authorize_app (app_id);
+        }
+      else
+        {
+          send_response (request, 2);
+
+          return;
+        }
+    }
+
+  if (!uri)
+    {
+      g_autofree char *path = NULL;
+
+      path = xdp_app_info_get_path_for_fd (request->app_info, fd, 0, NULL, NULL);
+      if (path == NULL)
+        {
+          /* Reject the request */
+          if (request->exported)
+            {
+              xdp_request_emit_response (XDP_REQUEST (request),
+                                         XDG_DESKTOP_PORTAL_RESPONSE_OTHER,
+                                         NULL);
+              request_unexport (request);
+            }
+          return;
+        }
+
+      basename = g_path_get_basename (path);
+
+      uri = g_filename_to_uri (path, NULL, NULL);
+      g_object_set_data_full (G_OBJECT (request), "uri", g_strdup (uri), g_free);
+    }
+
+  impl_request = xdp_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (impl)),
+                                                  G_DBUS_PROXY_FLAGS_NONE,
+                                                  g_dbus_proxy_get_name (G_DBUS_PROXY (impl)),
+                                                  request->id,
+                                                  NULL, &error);
+  request_set_impl_request (request, impl_request);
+
+  g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+  xdp_filter_options (options, &opt_builder,
+                      wallpaper_options, G_N_ELEMENTS (wallpaper_options),
+                      NULL);
```

Comment:
This will quietly drop the 'set-on' option if it is invalid. Does the backend handle its absence ok? What is the behavior? Might be good to document that.